README.md

```diff
@@ -1,167 +1,198 @@
+
+For question 2.5, you'll write one or more source code files and add them to the
+repository. You will then show the command used to run the files in the answer.
+Do _not_ upload any binary files for this section.
+
+Please do not edit anything outside the answers sections.
+
+
+# Q1: Mr Fussy Builds a Path
+
+# Q1
+
+Mr Fussy wants to build a path using a row of large square pavers. His path is
+one paver wide and an odd number of pavers long. Pavers come in three colors:
+red, green, blue.
+
+Mr Fussy has a rule: the sequence of colors must be symmetrical across the
+length of the path: the last paver must be the same color as the first, the
+second to last the same color as the second, and so on.
+
+This is a valid path:  rgbgr
+So is this: rrr
+
+This is not: rgb
+This is not: rggr    (it must be an odd length)
+
+## Q1.1  (5 points)
+
+Write a Chomsky type 2 grammar that describes any valid path containing at
+least one paver. Use `S` as the start state, and `r`, `g`, `b` as terminals that
+represent the tiles.
+
+## A1.1
+
+S -> r| g | b
+S -> r S r
+S-> g S g
+S -> b S b
+
+r -> "r"
+g -> "g"
+b -> "b" 
+
+## Q1.2  (1 point for the O() answer, 2 for the sentence)
+
+Using big-O notation, what is the likely memory requirement for a parser that
+can validate a particular path configuration, where `n` is the number of tiles?
+In one sentence, explain why.
+
+## A1.2
+
+It would be O(n) because the parser has to check every input to determine if it is valid.
+
+
+# Q2
+
+A simple sentence structure might be
+
+    "the" (zero or more adjectives) noun verb (optional adverb)
+
+Adjectives are "lazy" or "smelly"
+
+Nouns are "dog" or "cat"
+
+Verbs are "ate" and "ran"
+
+Adverbs are "slowly" and "noisily"
+
+The following are examples of valid sentences:
+
+* The smelly dog ran.
+* The smelly dog ran slowly.
+* The cat ate noisily.
+
+## Q2.1 (7 points)
+
+
+
+## A2.1
+
+<sentence> ::= "the" <optional_adj> <noun> <verb> <adverb>
+
+<adj> ::= "lazy" | "smelly" 
+<optional_adj> ::= <adJ> | <adj> <optional_adj> | ""
+
+<noun> ::= "dog" | "cat"
+
+<verb> ::= "ate" | "ran"
+
+<adverb> ::= "slowly" | "noisily" | ""
+
+
+## Q2.2 (5 points)
+
+Write this grammar using EBNF with common extensions
+
+## A2.2
+
+<sentence> ::= "the" {<adj>} <noun> <verb> [<adverb>]
+<adjective> ::= "lazy" | "smelly"
+<noun> ::= "dog" | "cat"
+<verb> ::= "ate" | "ran"
+<adverb> ::= "slowly" | "noisily"
+
+
+## Q2.3 (6 points)
+
+  Draw a diagram for an FSM which recognizes these sentences. Use EOI as the
+  event that occurs at the end-of-input. Start with a state named S0. Receiving
+  the word "the" in that state will transition to state S1. Name the final state
+  END.
+
+  You can hand draw it and snap a picture, or use a drawing tool. Either way,
+  upload the image, and put the file name in the answer below.
+
+  (Hint: my answer has seven states including the start and end states)
+
+
+## A2.3
+
+The file is in github. It is called State Machine.jpg

```

Comment:
# -2:

The diagram is correct except it doesn't deal with EOI: there's no way of knowing when you've successfully completed. For example, if the sentence was "the dog ate", you'd be in S3. Are we done? How could we know? Just because we ran out of input and we're still in the diagram doesn't mean the sentence is correct (for example the sentence "the dog" also finishes inside the diagram.

You need a DONE state, with EOI transitions from S3 and S4.